R/user-access.R

```diff
@@ -0,0 +1,39 @@
+#' Get information about when users last accessed a dataset
+#'
+#' Useful to get dataset-user interaction metadata about all of the users that
+#' have access to a particular dataset.
+#'
+#' @param dataset_entity a dataset object or a dataset entity (e.g. a single
+#' entry from `selectDatasetCatalog()`)
+#' @param user_fields metadata fields about users to include in the dataframe
+#' (default: c("name", "email", "access_time"))
+#'
+#' @return a dataframe with information about all users for a particular dataset
+#' @export
+userInfo <- function (dataset_entity, user_fields = c("name", "email", "access_time")) {
+    users_catalog <- ShojiCatalog(crGET(paste0(self(dataset_entity),"users/")))
+    all_users <- catalogToDataFrame(users_catalog, keys = user_fields)
+    all_users$dataset <- name(dataset_entity)
+    all_users$dataset_id <- id(dataset_entity)
+
+    return(all_users)
+}
+
+#' Get information about when users last any of the datasets in a project
+#'
+#' Useful to get dataset-user interaction metadata about all of the users that have
+#' access to all of the datasets in a project.
+#'
+#' @param project a character of the project name to gather metadata from
+#' @param ... additional arguments passed to [userInfo()] (e.g. user_fields)
+#'
+#' @return a dataframe with information about all users for a particular dataset
+#' @export
+projectUserInfo <- function (project, ...) {
+    sets <- selectDatasetCatalog(project = project)
+
+    list_of_dfs <- base::lapply(sets, userInfo, ...)
```

Comment:
Is `base::` required because `lapply` behaves differently on `sets`?

Reply:
If you're doing `base::` then something is wrong.

Reply:
I was trying to avoid the fact that by default `lapply` for ShojiCatalogs iterating over the index rather than the entities in the catalog themselves (cf https://github.com/Crunch-io/rcrunch/blob/master/R/shoji-catalog.R#L181). 

I did some digging, and it looks like that's not actually necessary (anymore?). Unless there's something that depends on it that's not tested, and I'm not seeing, we can remove that `lapply` definition without damaging anything.

Reply:
You don't want to load every dataset in the catalog. As currently implemented, loading a dataset eagerly fetches the variables catalog and variable order, and you don't want or need that. I could suggest a couple of more efficient ways to do that, but as I just commented below, I don't think that's something we should be doing at all.